feat(ssh_tunnel): APIs for SSH Tunnels

Antonio-RiveroMartnez opened the pull request:
### SUMMARY
Add our CREATE, UPDATE and DELETE APIs for SSH Tunnels.

- [X] CREATE API

> - [X] API endpoint calling the command
> - [X] Tests

- [X] UPDATE API
> - [X] API endpoint calling the command
> - [X] Tests

- [X] DELETE API
> - [X] API endpoint calling the command
> - [X] Tests


- [x] API + Schemas + Tests
    - [x] CREATE `POST /api/v1/database`
    - [x] UPDATE `UPDATE /api/v1/database/`
    - [x] DELETE `DELETE /api/v1/database/{db_id}/ssh_tunnel`



### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
<!--- Skip this if not applicable -->

### TESTING INSTRUCTIONS
<!--- Required! What steps can be taken to manually verify the changes? -->

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

Antonio-RiveroMartnez committed:
SSH Tunnel:

- Initial DELETE and CREATE ssh tunnel APIs
- Add happy path tests for each API

NOTE:
- Updating a DB might also result in creating a Tunnel, that flow will get tested in the UPDATE API changes.

Antonio-RiveroMartnez committed:
SSH Tunnel:

- Uncomment ssh tunnel creation for TestConnection command
- Update our tests
- Add update changes in API

Antonio-RiveroMartnez committed:
SSH Tunnel:

- Adjust our tests so it's clear our endpoint uses PATCH and not PUT
- Add delete cascading test: When deleting a database, the tunnel should be deleted

Antonio-RiveroMartnez committed:
SSH Tunnel:

- Add not found test when deleting the tunnel

Antonio-RiveroMartnez committed:
SSH Tunnel:

- FIx linting errors

Antonio-RiveroMartnez committed:
SSH Tunnel:

- Fix pre-commit errors

hughhhh commented on superset/databases/schemas.py:
```diff
@@ -365,6 +365,26 @@ class Meta:  # pylint: disable=too-few-public-methods
+    private_key_password = fields.String(required=False)
+
+    # remote binding port
+    bind_host = fields.String()
+    bind_port = fields.Integer()
```
we don't need to add this part to the schema

hughhhh commented on tests/integration_tests/databases/api_tests.py:
```diff
@@ -35,6 +35,8 @@
 
```
Can we add a test on `SSHTunnel` failure, to make sure we don't save the `Database` object as wel..

Antonio-RiveroMartnez committed:
SSH Tunnel:

- Remove bind properties from schema used in TestConnection
- Add test for SSH Tunnel failure and no DB creation

Antonio-RiveroMartnez committed:
SSH Tunnel:

- Make server_address, server_port and username required fields for our SSHTunnel schema
- Update the tests so we can consider new message with required fields

eschutho commented on superset/databases/api.py:
```diff
@@ -1204,3 +1210,58 @@ def validate_parameters(self) -> FlaskResponse:
+
+    @expose("/<int:pk>/ssh_tunnel/", methods=["DELETE"])
+    @protect()
+    @safe
```
IIRC, we no longer want to use this decorator because we want all errors to raise to be SIP41 compliant. DId I get that right @betodealmeida?

hughhhh replied:
+1 we should remove this so SIP-41 errors can be returned

eschutho commented on tests/integration_tests/databases/api_tests.py:
```diff
@@ -279,6 +281,262 @@ def test_create_database(self):
+            .filter(SSHTunnel.database_id == response.get("id"))
+            .one_or_none()
+        )
+        assert model_ssh_tunnel is None
```
nice!

eschutho commented on superset/databases/commands/update.py:
```diff
@@ -94,11 +95,35 @@ def run(self) -> Model:
+                    )
+                else:
+                    # We found an existing tunnel so we need to update it
+                    ssh_tunnel_model = SSHTunnelDAO.find_by_id(existing_ssh_tunnel.id)
```
wouldn't this lookup/instance here be the same as the existing_ssh_tunnel if it were not None?

Antonio-RiveroMartnez replied:
leftover from when it was a dict. Thanks!

eschutho commented on superset/databases/commands/update.py:
```diff
@@ -94,11 +95,35 @@ def run(self) -> Model:
             logger.exception(ex.exception)
             raise DatabaseUpdateFailedError() from ex
+        except DAOCreateFailedError as ex:
+            logger.exception(ex.exception)
```
the logger should log at the root api level

eschutho replied:
nit, but if the except block on both of these errors, is the same, you might as well just catch both errors in the same block, no?

hughhhh commented:
/testenv up

Antonio-RiveroMartnez committed:
SSH Tunnel:

- Changes from Code review feedback
- Solve conflict with base branch

Antonio-RiveroMartnez committed:
SSH Tunnel:

- Remove safe decorator so SIP-41 errors can be returned

Antonio-RiveroMartnez committed:
SSH Tunnel:

- Fix update tests

hughhhh committed:
debugging

Antonio-RiveroMartnez committed:
SSH Tunnel:

- Mask passwords for SSH when creating a new one
- Fix arg name in test connection

hughhhh committed:
refactor bind_host and bind_port

hughhhh commented on superset/databases/commands/update.py:
```diff
@@ -94,10 +95,31 @@ def run(self) -> Model:
+                existing_ssh_tunnel_model = DatabaseDAO.get_ssh_tunnel(database.id)
+                if existing_ssh_tunnel_model is None:
+                    # We couldn't found an existing tunnel so we need to create one
+                    ssh_tunnel = SSHTunnelDAO.create(
```
yea we should refactor these to use the commands instead, intially when we were writing these we prolly followed the pattern of the DatabaseDAO in this class

Antonio-RiveroMartnez committed:
SSH Tunnel:

- Use SSHTunnel commands instead of DAOs
- Update tests
- Fix linting and pre-commit

Antonio-RiveroMartnez committed:
SSH Tunnel:

- Flush only if creating a ssh tunnel

hughhhh commented on superset/models/core.py:
```diff
@@ -350,6 +350,19 @@ def set_sqlalchemy_uri(self, uri: str) -> None:
+            # Mask the real password with our password mask
+            self.ssh_tunnel[  # pylint: disable=no-member
+                "private_key_password"
+            ] = PASSWORD_MASK
```
this isn't overwriting the actual value of the password before saving right?

Antonio-RiveroMartnez replied:
Yeah, probably it's best to just getting rid of it in the response. Thanks!

Antonio-RiveroMartnez committed:
SSH Tunnel:

- Do not return passwords for SSH Tunnel in any API call

Antonio-RiveroMartnez committed:
SSH Tunnel:

- Using nested transactions so we can rollback if anything fails

Antonio-RiveroMartnez committed:
Merge branch 'create-sshtunnelconfig-tbl' into api_ssh_tunnel

Antonio-RiveroMartnez committed:
SSH Tunnel:

- Extend GET endpoint on Database APi so it includes SSH Tunnel if any

Antonio-RiveroMartnez committed:
Merge branch 'create-sshtunnelconfig-tbl' into api_ssh_tunnel

Antonio-RiveroMartnez committed:
Merge branch 'create-sshtunnelconfig-tbl' into api_ssh_tunnel

hughhhh commented on superset/databases/api.py:
```diff
@@ -219,6 +225,47 @@ class DatabaseRestApi(BaseSupersetModelRestApi):
+        ---
+        get:
+          description: >-
+            Get a database with its SSH Tunnel if any
```
Get a database

hughhhh commented on superset/databases/api.py:
```diff
@@ -280,6 +327,12 @@ def post(self) -> FlaskResponse:
 
+            # Return SSH Tunnel and hide passwords if any
+            if item.get("ssh_tunnel"):
+                item["ssh_tunnel"] = new_model.ssh_tunnel  # pylint: disable=no-member
```
create a util function that does the popping/deletion and returns the object without those values

Antonio-RiveroMartnez replied:
Sure! I'll indeed use our `mask` as I did here: https://github.com/apache/superset/pull/22513/files#diff-f9deb9b63f5f2db38b811ff092cf90b0d4d558223f743bbbb5b203c203d2d590R336 but sure, let me add that util. Thanks!

hughhhh commented on superset/databases/api.py:
```diff
@@ -361,6 +414,11 @@ def put(self, pk: int) -> Response:
                 item["parameters"] = changed_model.parameters
+            # Return SSH Tunnel and hide passwords if any
+            if item.get("ssh_tunnel"):
+                item["ssh_tunnel"] = changed_model.ssh_tunnel
```
use the util function here as well

hughhhh commented on superset/databases/commands/create.py:
```diff
@@ -71,17 +75,26 @@ def run(self) -> Model:
+                try:
+                    # So database.id is not None
+                    db.session.flush()
+                    CreateSSHTunnelCommand(database.id, ssh_tunnel_properties).run()
```
we need to have a reference for `ssh_tunnel` after it's created so it can pull down all the schema

https://github.com/apache/superset/pull/22199/files#diff-c3f1c5afb1b73f705f01bfa24cb4ca2ac66b84f82875abd3ddb73cb8147ea77eR100

Antonio-RiveroMartnez replied:
Good catch, let me add it there! Thanks

Antonio-RiveroMartnez committed:
SSH Tunnel:

- Add util method to mask password
- Adjust tests to consider password masking

Antonio-RiveroMartnez committed:
SSH Tunnel:

- Remove the extra description on GET endpoint for databases

Antonio-RiveroMartnez committed:
SSH Tunnel:

- Fix pre-commit error
